### PR TITLE
Replace serializers with callbacks

### PR DESCRIPTION
**Pull request by itspriddle**
**WIP - REVIEW ONLY**
## Serializers are bad m'kay

Serialized attributes in Rails are terribly inefficient. Each time they are checked for changes (for ActiveRecord::Dirty), the serializer's load/dump methods are called multiple times -- specifically for CryptKeeper this means encrypt/decrypt, and potentially several DB queries.

On a Rails 4.2 application with CryptKeeper master and the PostgresPgp encryptor, `Model.create` calls decrypt 3 times, and encrypt 8 times per encrypted attribute# `Model#save` calls decrypt 3 times, and encrypt 6 times per encrypted attribute.

For Rails 5, `Model.create` calls decrypt 2 times, and encrypt 3 times per attribute. `Model#save` calls decrypt 2 times and encrypt 4 times per attribute.

`Model.find` behaves as expected on Rails 4.2 and 5, with only one decrypt per encrypted attribute.

Additionally, checking dirty attributes via `Model#changes` will run several decrypt calls on _every_ encrypted attribute that was loaded. Eg: with `m = Model.select(:encrypted1, :encrypted2, :non_encrypted)`, `m.changes`, `m.encrypted1_changed?` or `m.encrypted2_changed?` will decrypt all encrypted fields. Checking `m.non_encrypted_changed?` does not.
## Possible fix for serializers

One idea I toyed around with was to cache the results of CryptKeepers serializers' dump/load calls. Basically something like this:

``` ruby
module Serializer
  @dumps = @loads = {}

  def dump(value)
    @dumps.fetch(value) do
      @dumps[value] = encrypt(value)
    end
  end

  def load(value)
    @loads.fetch(value) do
      @loads[value] = decrypt(value)
    end
  end
end
```

This does seem to work well, but I don't like the idea of caching these sensitive values in memory. Additionally, we would have to figure out a way to garbage clean older entries, otherwise the caches would continue to grow for the entire life of a Rails process.

Overall though, serlizers' inefficiencies are something we've been battling since Rails 3.2, and it doesn't seem to have improved much even with Rails 5.0.
## Why callbacks?

There are three main reasons I think callbacks are a better choice for CryptKeeper.
1. The most obvious reason, is that we can reduce encrypt/decrypt calls to, at most, 2 per attribute (one encrypt, one decrypt).
2. Dirty attribute handling is better because we avoid decrypt/encrypt calls when using `Model#changes` and friends. 
3. With callbacks we have full access to the model. This means we can eventually refactor further so that there are at most 1 encrypt and 1 decrypt query that would handle _all_ attributes. With a serializer this isn't possible since they only have access to the raw value and not the model/attribute name. If this PR is merged I have another work in progress to implement this.
## Why not callbacks?

FUD? There is a lot of hate for callbacks lately. It can be difficult to troubleshoot issues with them. However, I think that their usage in CryptKeeper is relatively simple, and I think the benefits outweigh any negatives when compared to serializers.

### CODE REVIEW

File: lib/crypt_keeper/callbacks.rb [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+    # Private: Decrypts the given value. If `CryptKeeper.stub_encryption` is
+    # enabled, or the given value is nil or a blank string, no decryption is
+    # run.
+    #
+    # value - The value to encrypt
+    #
+    # Returns a String or nil.
+    def decrypt(value)
+      if CryptKeeper.stub_encryption? || value.nil? || value == "".freeze
```
Comment (fabiokr) [inline lib/crypt_keeper/callbacks.rb]: Shouldn't `"".freeze` be memoized somewhere to be worth?

Comment (itspriddle) [inline lib/crypt_keeper/callbacks.rb]: I'm not sure if it matters in practice. The main reason I did this over just `Object#blank?` is that this is way more efficient. IIRC, 1 million iterations took `blank?` 12 seconds vs ~1.75 to check nil/"". But I will run some benchmarks to see if storing the frozen string somewhere makes a difference.

**Comment by itspriddle** [comment]
Some benchmarks regarding the change from `String#blank?` in encrypt/decrypt methods.

tl;dr: Checking with `.nil?` and `string == "".frozen"` is ~8 times faster than using `Object#blank?`. This won't matter as much with callbacks reducing the overall number of encrypt/decrypt calls, but its a low hanging fruit we should go after.

### Test 1, Checking a string value using an empty frozen string (`"".frozen`)

```
require "benchmark/ips"
gem "activesupport", "4.2.6"
require "active_support/core_ext/string"

FROZEN = "".freeze

string = ""

Benchmark.ips do |x|
  x.report "String#blank?" do
    string.blank?
  end

  x.report "String#nil? || String#== ''.freeze" do
    string.nil? || string == FROZEN
  end

  x.compare!
end
```

Result:

```
ruby bench.rb
Warming up --------------------------------------
       String#blank?    54.783k i/100ms
String#nil? || String#== ''.freeze
                       212.411k i/100ms
Calculating -------------------------------------
       String#blank?    879.919k (± 5.9%) i/s -      4.437M in   5.061889s
String#nil? || String#== ''.freeze
                          6.850M (± 6.5%) i/s -     34.198M in   5.014137s

Comparison:
String#nil? || String#== ''.freeze:  6850459.7 i/s
       String#blank?:   879918.9 i/s - 7.79x slower

ruby bench.rb  14.26s user 0.10s system 99% cpu 14.420 total
```

---

### Test 2, Checking a string value using an empty non-frozen string (`""`)

```
ruby bench.rb
Warming up --------------------------------------
       String#blank?    51.233k i/100ms
String#nil? || String#== ''
                       182.909k i/100ms
Calculating -------------------------------------
       String#blank?    884.049k (± 6.7%) i/s -      4.406M in   5.007717s
String#nil? || String#== ''
                          4.635M (± 7.0%) i/s -     23.047M in   5.000579s

Comparison:
String#nil? || String#== '':  4634518.7 i/s
       String#blank?:   884049.1 i/s - 5.24x slower

ruby bench.rb  14.16s user 0.10s system 99% cpu 14.341 total
```

---

### Test 3, checking a nil value

```
require "benchmark/ips"
gem "activesupport", "4.2.6"
require "active_support/core_ext/string"

FROZEN = "".freeze

string = nil

Benchmark.ips do |x|
  x.report "String#blank?" do
    string.blank?
  end

  x.report "String#nil? || String#== ''.freeze" do
    string.nil? || string == FROZEN
  end

  x.compare!
end
```

Result:

```
ruby bench.rb
Warming up --------------------------------------
       String#blank?   222.780k i/100ms
String#nil? || String#== ''.freeze
                       232.980k i/100ms
Calculating -------------------------------------
       String#blank?      7.950M (± 7.4%) i/s -     39.655M in   5.017970s
String#nil? || String#== ''.freeze
                          7.962M (± 8.5%) i/s -     39.607M in   5.010767s

Comparison:
String#nil? || String#== ''.freeze:  7961549.0 i/s
       String#blank?:  7949684.7 i/s - same-ish: difference falls within error

ruby bench.rb  14.22s user 0.07s system 99% cpu 14.377 total
```

File: lib/crypt_keeper/callbacks.rb (hunk 1 above)
Comment (itspriddle) [inline lib/crypt_keeper/callbacks.rb]: I ran some benchmarks and the memoized version is actually slightly faster, so I will end up implementing that change.

**Comment by tfwright** [comment]
Interesting approach. I think that the main drawback to using callbacks for this is that it's not what they seem to be intended for.  They are supposed to define functions to be run _on_ the instance, not _define_ the instance. Unless I'm mistaken, this will break, for example, a model that _prepends_ a callback that expects access to the encrypted data. 

That said, when working on CK I noticed how many, _many_ times encrypt/decrypt were being called so the tradeoffs might be worth the performance gains. But I don't think it's the optimum approach from a pure design perspective.

**Comment by itspriddle** [comment]
> Interesting approach. I think that the main drawback to using callbacks for this is that it's not what they seem to be intended for. They are supposed to define functions to be run on the instance, not define the instance. Unless I'm mistaken, this will break, for example, a model that prepends a callback that expects access to the encrypted data.
> 
> That said, when working on CK I noticed how many, many times encrypt/decrypt were being called so the tradeoffs might be worth the performance gains. But I don't think it's the optimum approach from a pure design perspective.

@tfwright Yeah, if someone used a `before_save :blah, prepend: true` that needed access to encrypted data it would not work. Unless I'm missing something, though, that same thing would be an issue with serializers since the attributes would still be decrypted when the callback fires.

If performance weren't a concern here I would also prefer serializers. I am open to _any_ alternatives. We could even try the caching I mentioned above (although we still wouldn't be able to consolidate encrypt/decrypts for multiple attributes into one query).
